src/libsystemd/sd-event/sd-event.c

```diff
@@ -1241,6 +1241,16 @@ _public_ int sd_event_add_child(
         assert_return(e->state != SD_EVENT_FINISHED, -ESTALE);
         assert_return(!event_pid_changed(e), -ECHILD);
 
+        if (e->n_enabled_child_sources == 0) {
+                /* Caller must block SIGCHLD before using us to watch children, even if pidfd is available,
+                 * for compatibility with pre-pidfd */
+                r = signal_is_blocked(SIGCHLD);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        return -EBUSY;
+        }
```

Comment:
Well, they still really have to turn SIGCHLD off then, because otherwise the default behaviour of SIGCHLD is to reap the child, but we don't want that, we still want to reap it ourselves with waitid() so that we can retrieve the exit status and such. Hence, blocking SIGCHLD is not an option either way, it's required for correct behaviour.